verify M2 and Ivy patterns contain the required tokens in any pattern

aSemy opened the pull request:
Permit more flexible repository layout patterns.

Instead of strictly checking for an exact repository pattern, instead verify the patterns contain the required tokens.

- M2 patterns must contain: `organisation`, `module`, `artifact`.
- Ivy patterns must contain: `artifact`.

The existing checks already required these tokens. They are still required, but the position is not fixed.

Signed-off-by: Adam <897017+aSemy@users.noreply.github.com>

I've tested this against the example project in #33674 and it works.

Fix #33674

<!--- The issue this PR addresses -->
<!-- Fixes #? -->

### Context
<!--- Why do you believe many users will benefit from this change? -->
<!--- Link to relevant issues or forum discussions here -->

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] ~Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.~ Not possible due to https://github.com/gradle/gradle/issues/35063
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

big-guy commented:
We're not going to be able to include this in 9.0.  We would need to investigate how this would impact publishing so that we do not publish unresolvable GMM.

aSemy committed:
verify M2 and Ivy patterns contain the required tokens in any pattern

Permit more flexible repository layout patterns.

Instead of strictly checking for an exact repository pattern, instead verify the patterns contain the required tokens.

- M2 patterns must contain: `organisation`, `module`, `artifact`.
- Ivy patterns must contain: `artifact`.

The existing checks already required these tokens. They are still required, but the position is not fixed.

Signed-off-by: Adam <897017+aSemy@users.noreply.github.com>

hegyibalint commented:
Thank you for your proposed contribution!

This PR has a valid DCO and tests. The relevant team for this area will confirm the design of the implementation choices.

---

@jvandort please check the labels and other related labels if necessary.

aSemy commented:
> Do you need to set M2compatible in the reproducer for #33674?

iiuc it's not about whether M2compatible is enabled or disabled on the user-defined repo. Gradle will always use M2ResourcePattern to resolve artifacts via Gradle module metadata.
